### PR TITLE
GET-56 Build Job Profile Search page and Results page

### PR DESCRIPTION
**Pull request by cwrw**
jira-ticket: https://dfedigital.atlassian.net/browse/GET-56

* Build Job Profile responsive search and results page 
* Search is case insensitive
* Add base for beta banner and breadcrumbs

### CODE REVIEW

File: spec/models/job_profile_spec.rb [hunk 1]
```diff
@@ -30,4 +30,24 @@
...
+      job_profile = create(:job_profile, name: 'Beverage Dissemination Officer')
+
+      expect(described_class.search('Dissemination')).to contain_exactly(job_profile)
+    end
+
+    it 'returns nothing if no job profile is matched' do
+      create(:job_profile, name: 'Dream Alchemist')
+
+      expect(described_class.search('Beverage Dissemination Officer')).to be_empty
+    end
```
Comment (cwrw) [inline spec/models/job_profile_spec.rb]: add extra test for `nil` name

File: app/models/category.rb [hunk 2]
```diff
@@ -1,4 +1,9 @@
 class Category < ApplicationRecord
   has_many :job_profile_categories
   has_many :job_profiles, through: :job_profile_categories, inverse_of: :categories
+
+  def self.with_job_profiles
+    includes(:job_profile_categories)
+      .where.not(job_profile_categories: { job_profile_id: nil })
```
Comment (cwrw) [inline app/models/category.rb]: new line at `.not`

File: app/webpacker/styles/_job-profile.scss [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    color: govuk-colour("white")
+  }
+}
+.search-button {
+  border: 0;
+  width:40px;
+  height:40px;
+  float: left;
+  background-color: govuk-colour("black");
+  background-image: url("../images/search_icon.svg");
```
Comment (cwrw) [inline app/webpacker/styles/_job-profile.scss]: add assets url instead

File: app/models/job_profile.rb [hunk 4]
```diff
@@ -3,4 +3,8 @@ class JobProfile < ApplicationRecord
   has_many :job_profile_skills
   has_many :categories, through: :job_profile_categories, inverse_of: :job_profiles
   has_many :skills, through: :job_profile_skills, inverse_of: :job_profiles
+
+  def self.search(name)
+    where('name LIKE ?', "%#{name}%") if name
```
Comment (smoothcontract) [inline app/models/job_profile.rb]: Maybe use ILIKE (case independent) ?

File: app/controllers/job_profiles_controller.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  end
+
+  def index
+    @job_profiles = JobProfile.search(job_profile_params[:name])
+  end
+
+  private
+
+  def job_profile_params
+    params.permit(:name)
```
Comment (cwrw) [inline app/controllers/job_profiles_controller.rb]: add require `.require(:job_profile)`

File: app/models/job_profile.rb (hunk 4 above)
Comment (smoothcontract) [reply app/models/job_profile.rb]: Alternatively there’s a postgres column type that supports case independent comparison operators by default - we could swap to that. Would have to google that, can’t recall off the top of my head.

File: spec/models/category_spec.rb [hunk 6]
```diff
@@ -13,4 +13,30 @@
...
+    end
+
+    it 'does not return categories not linked to job profiles' do
+      category = create(:category, job_profiles: [create(:job_profile)])
+      create(:category)
+      expect(Category.with_job_profiles).to contain_exactly(category)
+    end
+
+    it 'returns nothing if no categories linked to job profiles' do
+      create_list(:category, 4)
```
Comment (smoothcontract) [inline spec/models/category_spec.rb]: Maybe just create 2 categories here? That’s enough to prove the spec without slowing it unnecessarily.

File: app/models/job_profile.rb (hunk 4 above)
Comment (smoothcontract) [reply app/models/job_profile.rb]: See https://nandovieira.com/using-insensitive-case-columns-in-postgresql-with-citext

Comment (cwrw) [reply app/models/job_profile.rb]: I've went with `ILIKE` for now, I think we need to think generally about search and what we would like it's behaviour to be like before committing to anything past queries

File: app/views/shared/_contact_us.html.erb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<div class="govuk-grid-column-one-third">
+  <div class="govuk-related-items govuk-!-padding-bottom-4 top-border__container">
+    <h2 class="govuk-heading-m"><%= t('contact_us.title') %></h2>
+    <p class="govuk-body-m">
+      <%= t('contact_us.body') %> <a href="tel:+442079476330" <%= t('contact_us.telephone_link') %> class="govuk-!-padding-left-1 govuk-link"><%= t('contact_us.telephone') %></a>
```
Comment (cwrw) [inline app/views/shared/_contact_us.html.erb]: fix href

File: spec/models/job_profile_spec.rb [hunk 8]
```diff
@@ -30,4 +30,60 @@
       expect(recommended_job.skills).to match [skill]
     end
   end
+
+  describe '.search' do
+    it 'returns a job profile if a name matches exactly' do
+      job_profile = create(:job_profile, name: 'Beverage Dissemination Officer')
```
Comment (tavy88) [inline spec/models/job_profile_spec.rb]: could we use `let(:job_profile)/let!` ?

File: spec/models/job_profile_spec.rb [hunk 9]
```diff
@@ -30,4 +30,60 @@
...
+
+      expect(described_class.search('Beverage Dissemination Officer')).to contain_exactly(job_profile)
+    end
+
+    it 'returns a job profile if a name is like an existing job profile name' do
+      job_profile = create(:job_profile, name: 'Beverage Dissemination Officer')
```
Comment (tavy88) [inline spec/models/job_profile_spec.rb]: could we use let(:job_profile)/let! ?

File: spec/models/job_profile_spec.rb (hunk 8 above)
Comment (cwrw) [reply spec/models/job_profile_spec.rb]: I think there are two approaches, sharing objects across tests vs. creating tests which are self contained. I am following the trend of self contained tests where we create everything in there. If there is something bigger or more intricate we can move it out to a let

File: spec/models/job_profile_spec.rb (hunk 9 above)
Comment (cwrw) [reply spec/models/job_profile_spec.rb]: same above

File: app/views/job_profiles/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    <%= link_to(t('breadcrumb.job_profiles_search'), search_job_profiles_path, class: 'govuk-breadcrumbs__link') %>
+  </li>
+  <li class="govuk-breadcrumbs__list-item"><%= t('breadcrumb.job_profiles_result') %></li>
+<% end %>
+
+<div class="govuk-grid-row govuk-!-padding-top-7">
+  <div class="govuk-grid-column-two-thirds" >
+    <h3 class="govuk-heading-l"><%= t('.title') %></h1>
+    <div class="govuk-grid-column-full govuk-!-padding-0">
+      <%= form_tag job_profiles_path, method: 'get' do %>
```
[the remark this answers is not in the document]
Comment (cwrw) [reply app/views/job_profiles/index.html.erb]: I think for now it's best to keep them separate as this form is different across the index/search pages regarding styles.

File: app/controllers/job_profiles_controller.rb [hunk 11]
```diff
@@ -0,0 +1,15 @@
+class JobProfilesController < ApplicationController
+  def search
+    @job_categories = Category.with_job_profiles
+  end
```
[the remark this answers is not in the document]
Comment (cwrw) [reply app/controllers/job_profiles_controller.rb]: This was added initially as we thought we'd be including a list of categories on this page, I'll remove as it's not needed on the page anymore

File: app/controllers/job_profiles_controller.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+  def index
+    @job_profiles = JobProfile.search(job_profile_params[:name])
+  end
+
+  private
+
+  def job_profile_params
+    params.require(:job_profile).permit(:name)
+  end
```
[the remark this answers is not in the document]
Comment (cwrw) [reply app/controllers/job_profiles_controller.rb]: yes, we want to only allow name for now

File: app/models/category.rb [hunk 13]
```diff
@@ -1,4 +1,10 @@
 class Category < ApplicationRecord
   has_many :job_profile_categories
   has_many :job_profiles, through: :job_profile_categories, inverse_of: :categories
+
+  def self.with_job_profiles
+    includes(:job_profile_categories)
+      .where
+      .not(job_profile_categories: { job_profile_id: nil })
+  end
```
[the remark this answers is not in the document]
Comment (cwrw) [reply app/models/category.rb]: I think for now it's safe to leave as is until we get to work on this in the job category ticket, as it seems we don't need this anymore in this ticket

File: app/controllers/job_profiles_controller.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (tavy88) [reply app/controllers/job_profiles_controller.rb]: 👍

File: app/models/category.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (tavy88) [reply app/models/category.rb]: 👍

File: app/views/job_profiles/index.html.erb (hunk 10 above)
[the remark this answers is not in the document]
Comment (tavy88) [reply app/views/job_profiles/index.html.erb]: I doesn't need to be a shared partial, but it's more about drying this index page.

File: spec/models/job_profile_spec.rb (hunk 8 above)
Comment (tavy88) [reply spec/models/job_profile_spec.rb]: Ok, but using a let would dry your test steps as you are calling create multiple times.

File: app/views/job_profiles/index.html.erb (hunk 10 above)
[the remark this answers is not in the document]
Comment (cwrw) [reply app/views/job_profiles/index.html.erb]: I think both the search and index page are quite small, and not worth extracting partials out right now (unless they are reused then that would be a good case to start). If we notice the pages becoming unwieldy we can start doing that.

File: app/views/job_profiles/search.html.erb [hunk 14]
```diff
@@ -0,0 +1,22 @@
+<% content_for :breadcrumb do %>
+  <li class="govuk-breadcrumbs__list-item">
+    <%= link_to(t('breadcrumb.home'), root_path, class: 'govuk-breadcrumbs__link') %>
+  </li>
+  <li class="govuk-breadcrumbs__list-item"><%= t('breadcrumb.job_profiles_search') %></li>
+<% end %>
+
```
[the remark this answers is not in the document]
Comment (cwrw) [reply app/views/job_profiles/search.html.erb]: same as above

File: app/views/job_profiles/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,40 @@
+<% content_for :breadcrumb do %>
+  <li class="govuk-breadcrumbs__list-item">
+    <%= link_to(t('breadcrumb.home'), root_path, class: 'govuk-breadcrumbs__link') %>
+  </li>
+  <li class="govuk-breadcrumbs__list-item">
+    <%= link_to(t('breadcrumb.job_profiles_search'), search_job_profiles_path, class: 'govuk-breadcrumbs__link') %>
+  </li>
+  <li class="govuk-breadcrumbs__list-item"><%= t('breadcrumb.job_profiles_result') %></li>
+<% end %>
+
```
[the remark this answers is not in the document]
Comment (cwrw) [reply app/views/job_profiles/index.html.erb]: I think breadcrumbs is a good candidate for extraction to shared partials. It would be good to see a pattern emerge first (further from 2 views) so we can build a helper for those. For now we can keep the same until then
